### PR TITLE
[Subtitles][Webvtt] Implemented WebVTT parser

### PR DESCRIPTION
**Pull request by CastagnaIT**
## Description

Implement the WebVTT parser to allow use WebVTT external subtitles `*.vtt` files and streams from InputStream addon

~It depends from PR #20308~

This code follow W3C standard https://www.w3.org/TR/webvtt1/
Almost all features are implemented including partially the CSS parsing,
due to some Libass rendering limits some particular feature are not implemented like Cue region,
in any case interested parties are documented in the code.

Comparing with VLC, this parser has a more complete overall handling, and also handles better some parsing problems caused by malformed files, where VLC fails.

There are some things to be noted that are not supported:
- Cue region (text will be displayed at the bottom)
- Cue "line" setting as number value
- Vertical text (used for some specific asian languages only)

With regard to `Cue region` and `Cue "line" setting as number value` i have not found any information on where are used,
sample files are also very rare to find, it is possible that they are only applied only on some specifics web videos content.
It might be possible to implement some kind of support for these functionalities in the future although it is not easy to do so.
For example VLC have a limited supporto to `Cue region` and `Cue "line" setting as number value` which subtitles are not always displayed correctly on screen.

## Motivation and context

This is a first start of a bigger project that I would like, time permitting, to complete, which also concerns InputStream Adaptive,
and i hope with the help of others devs. This will mainly provide all VOD addons a full subtitle support.

Currently WEBVTT subtitles are parsed in a bad way by InputStream Adaptive (and also TTML subtitles),
ISA has no real subtitle parser, mainly cuts out all metadata and keeps only the text,
this works partially, but it destroys all the features of these subtitles formats, especially when subtitles are strategically placed for discussions or multilanguage purpose.

**How work currently:**
ISA "parse" and "convert" WebVTT type in to SubRip type --> feed the converted stream to Kodi --> Kodi convert again the SubRip stream to Ass --> rendering
**How (will) works now (after merged https://github.com/xbmc/inputstream.adaptive/pull/820):**
ISA feed the stream to Kodi --> Kodi WebVTT parser handle the stream --> rendering

This concerns the format WebVTT but later it will be necessary to do the same with the TTML format type.

**Different WebVTT formats handled**
There are two types of WebVTT formats, both handled by ISA (now by Kodi):
- "Text" format type, handled by this parser
- "ISOBMFF" format type, is MP4 specifics, was implemented in ISA in a limited way without many code checks.

**Embedded formats**
MKV video files support WebVTT format types, unfortunately we are tied to FFMPEG and FFMPEG is very late in implementing this support (ATM support the oldest format only, is not same thing), so we have to wait for their new implementation, then it will be automatically supported by Kodi without any effort.

## How has this been tested?

* External files
I have tested with the following test files, some modified for the purpose to test robustness of the code
[WEBVTT test files.zip](https://github.com/xbmc/xbmc/files/7279478/WEBVTT.test.files.zip)

* Streams (InputStream)
There are three type of streams:
-Webvtt "as single file" with Net***x addon
-Webvtt segmented with Dis€y+ addon
-Webvtt fMP4 data (ISOBMFF) with a sample mp4 stream

## What is the effect on users?

Displaying subtitles in the most appropriate way

## Review hint

The file:

xbmc/utils/Mime.cpp

the spaces have been fixed to make jenkins happy, so the changes are not immediately apparent,
to check the changes on this file follow this link:
https://github.com/CastagnaIT/xbmc/commit/39d3fd3669ab28f8d0113bd3e642e2c2b1613fb3?w=1

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/VideoPlayer/DVDSubtitles/WebVTTHandler.cpp
Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDSubtitles/WebVTTHandler.cpp:600]: Style: Can you right-align comments to 80 chars? This helps readability.

File: xbmc/cores/VideoPlayer/DVDSubtitles/WebVTTHandler.cpp
Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDSubtitles/WebVTTHandler.cpp:604]: Grammar: Change tense of verb `check` to `checking`

File: xbmc/cores/VideoPlayer/DVDSubtitles/WebVTTHandler.cpp
Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDSubtitles/WebVTTHandler.cpp:662]: Grammar: Change `not works` to `doesn't work`

File: xbmc/utils/CharsetConverter.h
Comment (garbear) [inline xbmc/utils/CharsetConverter.h:148]: Grammar: Change `have` to `has`

**Comment by garbear** [comment]
Small request:

Can you (preferably) split the 500 lines of whitespace changes into a cosmetic commit that comes first in the PR to aid reviewability?

Alternatively, if it's not worth the effort, just stick a link to the commit URL including `?w=1` in the PR description for reviewability so that ppl have a link to the commit without whitespace formatting. (The URL would look like `https://github.com/CastagnaIT/xbmc/commit/a91a8b7988cf834bb7de08d4b5204f6b882d1ac9?w=1`).

**Review by garbear — APPROVED** [review]
The code is clear and well-documented. Thanks for the feature!

I know it's only a draft PR, but I'm leaving a green check here to show that I approve the PR and my first pass has found only style and grammar changes. Please hold off on merging until I leave another check mark after draft status is removed.

File: xbmc/cores/VideoPlayer/DVDSubtitles/WebVTTHandler.cpp
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/DVDSubtitles/WebVTTHandler.cpp:410]: is not intended for production, i hope to implement this before make the PR ready, when the other subtitle PR will be merged

Comment (garbear) [reply xbmc/cores/VideoPlayer/DVDSubtitles/WebVTTHandler.cpp]: OK then good to use your style so that it gets addressed before merge

**Comment by scott967** [comment]
I reviewed WebVTTHandler from a requirements view.  Seems reasonable but I have not seen complex cue styles so hard to know what the user expectation is.  I agree that the "line" parameter as an integer seems poorly defined in spec; what I have seen is -1 for bottom and 0 for top.  From what I can see, embedded WebVTT might be an issue, as it doesn't seem like encoders are robust, so external WebVTT files probably are the only realistic option for local playback with VTT features.

**Review by enen92 — APPROVED** [review]
Can't spot anything wrong with the code, untested but I guess it won't affect current functionality as this is a new parser/feature

**Pull request merged**
